File: Q11.m
w = [0:1:500]*pi/500
x = [1,-0.5,-0.3,-0.1]
y = zeros(1,length(w))
for i =1 : length(w)
   for n=-1 :2
       y(i) = y(i) + x(n+2)*exp(-j*w(i)*n);

   end
end

subplot (2,2,1);
plot(w/pi, real(y),"m");
xlabel("f");
ylabel('real');
subplot (2,2,2);
plot(w/pi, angle(y),"m");
xlabel("f");
ylabel ("Angel");
subplot(2,2,3);
plot(w/pi, imag(y),"m");
xlabel("f");
ylabel ("imaginary");
subplot(2,2,4);
plot (w/pi,abs(y),"m");
xlabel("f");
ylabel("Amplitude")

File: Q11.py
import numpy as np
import matplotlib.pyplot as plt
import os
import random
import scipy as scp
from scipy.signal import lfilter, lfilter_zi

# Define the sequence x[n]
x = np.array([1, -0.5, -0.3, -0.1])

# Compute the discrete-time Fourier transform of x[n] at 501 equispaced frequencies between [0,π]
frequencies = np.linspace(0, np.pi, 501)
X = np.fft.fft(x, len(frequencies))

print(X)
##

# Plot the magnitude of X(e^jw)
plt.figure()
plt.plot(frequencies, np.abs(X))
plt.xlabel('Frequency (rad/sample)')
plt.ylabel('Magnitude')
plt.title('Magnitude of X(e^jw)')

# Plot the angle of X(e^jw)
plt.figure()
plt.plot(frequencies, np.angle(X))
plt.xlabel('Frequency (rad/sample)')
plt.ylabel('Angle (rad)')
plt.title('Angle of X(e^jw)')
plt.show()

# Plot the real part of X(e^jw)
plt.figure()
plt.plot(frequencies, np.real(X))
plt.xlabel('Frequency (rad/sample)')
plt.ylabel('Real')
plt.title('Real part of X(e^jw)')
plt.show()

# Plot the imaginary part of X(e^jw)
plt.figure()
plt.plot(frequencies, np.imag(X))
plt.xlabel('Frequency (rad/sample)')
plt.ylabel('Imaginary')
plt.title('Imaginary part of X(e^jw)')
plt.show()